pkg/controller/projectreference/projectreference_adapter.go

```diff
@@ -0,0 +1,343 @@
+	"us-west2":                true,
+}
+
+//ReferenceAdapater is used to do all the processing of the ProjectReference type inside the reconcile loop
+type ReferenceAdapater struct {
+	projectClaim     *gcpv1alpha1.ProjectClaim
+	projectReference *gcpv1alpha1.ProjectReference
+	logger           logr.Logger
+	kubeClient       client.Client
+	gcpClient        gcpclient.Client
+}
+
+func newReferenceAdapater(projectReference *gcpv1alpha1.ProjectReference, logger logr.Logger, client client.Client, gcpClient gcpclient.Client) (*ReferenceAdapater, error) {
+	projectClaim, err := getMatchingClaimLink(projectReference, client)
+	if err != nil {
+		return &ReferenceAdapater{}, err
+	}
+	return &ReferenceAdapater{
+		projectClaim:     projectClaim,
+		projectReference: projectReference,
+		logger:           logger,
+		kubeClient:       client,
+		gcpClient:        gcpClient,
+	}, nil
+}
+
+func getMatchingClaimLink(projectReference *gcpv1alpha1.ProjectReference, client client.Client) (*gcpv1alpha1.ProjectClaim, error) {
+	projectClaim := &gcpv1alpha1.ProjectClaim{}
+	err := client.Get(context.TODO(), types.NamespacedName{Name: projectReference.Spec.ProjectClaimCRLink.Name, Namespace: projectReference.Spec.ProjectClaimCRLink.Namespace}, projectClaim)
+	if err != nil {
+		return &gcpv1alpha1.ProjectClaim{}, err
+
+	}
+	return projectClaim, nil
+}
+
+// updateProjectID updates the ProjectReference with a unique ID for the ProjectID
+func (r *ReferenceAdapater) updateProjectID() error {
+	guid := uuid.New().String()
+	hashing := sha1.New()
+	hashing.Write([]byte(guid))
+	uuidsum := fmt.Sprintf("%x", hashing.Sum(nil))
+	shortuuid := uuidsum[0:26]
+	r.projectReference.Spec.GCPProjectID = "osd-" + shortuuid
+	return r.kubeClient.Update(context.TODO(), r.projectReference)
+}
+
+// updateProjectID updates the ProjectReference with a unique ID for the ProjectID
+func (r *ReferenceAdapater) clearProjectID() error {
+	r.projectReference.Spec.GCPProjectID = ""
+	return r.kubeClient.Update(context.TODO(), r.projectReference)
+}
+
+// checkRequirements checks that region is supported
+func (r *ReferenceAdapater) checkRequirements() error {
+	if _, ok := supportedRegions[r.projectClaim.Spec.Region]; !ok {
+		return operrors.ErrRegionNotSupported
+	}
+	return nil
+}
+
+func (r *ReferenceAdapater) createProject(parentFolderID string) error {
+	// Get existing projects
+	projects, err := r.gcpClient.ListProjects()
+	if err != nil {
+		return err
+	}
+
+	projectMap := convertProjectsToMap(projects)
+
+	if project, ok := projectMap[r.projectReference.Spec.GCPProjectID]; ok {
+		switch project.LifecycleState {
+		case "ACTIVE":
+			r.logger.Info("Project lifecycleState == ACTIVE")
+			return nil
+		case "DELETE_REQUESTED":
+			return operrors.ErrInactiveProject
+		default:
+			r.logger.Error(operrors.ErrUnexpectedLifecycleState, "Unexpected LifecycleState", project.LifecycleState)
+			return operrors.ErrUnexpectedLifecycleState
+
+		}
+		return nil
```

Comment:
It is unreachable code because you have individual `return` statement for each case.

```suggestion

```